File: examples/clients/python/blink_client.py
#!/usr/bin/env python3
"""Example Blink-Store client (TCP). GET/SET/DELETE/USAGE/QUIT, VALUE is base64.
Usage: python blink_client.py [host [port]]"""

import base64
import socket
import sys

def main():
    host = sys.argv[1] if len(sys.argv) > 1 else "127.0.0.1"
    port = int(sys.argv[2]) if len(sys.argv) > 2 else 8765
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.connect((host, port))
        reader = sock.makefile("r", encoding="utf-8", newline="\n")
        writer = sock.makefile("w", encoding="utf-8", newline="\n")
        print("Commands: GET <key> | SET <key> <value> | DELETE <key> | USAGE | QUIT", flush=True)
        try:
            while True:
                line = input().strip()
                if not line or line.upper() == "QUIT":
                    break
                writer.write(line + "\n")
                writer.flush()
                resp = reader.readline()
                if not resp:
                    break
                resp = resp.rstrip("\n")
                if resp.startswith("VALUE "):
                    try:
                        out = base64.b64decode(resp[6:].strip()).decode("utf-8", errors="replace")
                        print(out)
                    except Exception:
                        print(resp)
                else:
                    print(resp)
        finally:
            try:
                writer.write("QUIT\n")
                writer.flush()
            except Exception:
                pass

if __name__ == "__main__":
    main()

File: examples/clients/node/blink_client.js
#!/usr/bin/env node
/** Example Blink Store client (TCP). Protocol: GET/SET/DELETE/USAGE/QUIT, VALUE is base64.
 *
 * Usage:
 *   node blink_client.js [host [port]]
 *   node blink_client.js 127.0.0.1 8765
 */

const net = require('net');
const readline = require('readline');

const host = process.argv[2] || '127.0.0.1';
const port = parseInt(process.argv[3], 10) || 8765;

const sock = net.createConnection(port, host, () => {});
const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

let buffer = '';

sock.on('data', (data) => {
  buffer += data.toString();
  let idx;
  while ((idx = buffer.indexOf('\n')) !== -1) {
    const line = buffer.slice(0, idx).trim();
    buffer = buffer.slice(idx + 1);
    if (line.startsWith('VALUE ')) {
      const b64 = line.slice(6).trim();
      try {
        console.log(Buffer.from(b64, 'base64').toString('utf8'));
      } catch {
        console.log(line);
      }
    } else {
      console.log(line);
    }
  }
});

sock.on('close', () => process.exit(0));
sock.on('error', (err) => { console.error(err); process.exit(1); });

console.log('Commands: GET <key> | SET <key> <value> | DELETE <key> | USAGE | QUIT');

rl.on('line', (line) => {
  line = line.trim();
  if (!line) return;
  if (line.toUpperCase() === 'QUIT') {
    sock.write('QUIT\n');
    sock.end();
    return;
  }
  sock.write(line + '\n');
});
